src/lustre/lustreTypeChecker.ml

```diff
@@ -0,0 +1,1379 @@
+  | Mode (pos, name, _, _) -> R.ok (add_ty ctx name (Bool pos)) 
+  | ContractCall _ -> R.ok ctx
+  
+and check_type_contract_decl: tc_context -> LA.contract_node_decl -> unit tc_result
+  = fun ctx (cname, params, args, rets, contract) ->
+  Log.log L_debug "TC Contract decl: %a {" LA.pp_print_ident cname 
+  (* build the appropriate local context *)
+  ; let arg_ctx = List.fold_left union ctx (List.map extract_arg_ctx args) in
+  let ret_ctx = List.fold_left union arg_ctx (List.map extract_ret_ctx rets) in
+  let local_const_ctx = List.fold_left union ret_ctx (List.map extract_consts args) in
+  (* get the local const var declarations into the context *)
+  R.seq (List.map (extract_contract_eqn_context local_const_ctx) contract)
+  >>= fun ctxs ->
+  let local_ctx = List.fold_left union local_const_ctx ctxs in
+  check_type_contract local_ctx contract >>= fun _ -> 
+  R.ok (Log.log L_debug "TC Contract Decl %a done }" LA.pp_print_ident cname)
+
+and check_type_contract: tc_context -> LA.contract -> unit tc_result
+  = fun ctx eqns -> R.seq_ (List.map (check_contract_node_eqn ctx) eqns)
+
+and check_contract_node_eqn: tc_context -> LA.contract_node_equation -> unit tc_result
+  = fun ctx eqn ->
+  Log.log L_debug "checking contract with equation %a" LA.pp_print_contract_item eqn 
+  ; match eqn with
+  | GhostConst _
+  | GhostVar _ ->  R.ok () (* These is already checked while extracting ctx *)
+  | Assume (pos, _, _, e) -> check_type_expr ctx e (Bool pos)
+  | Guarantee (pos, _, _, e) -> check_type_expr ctx e (Bool pos)
+  | Mode (pos, _, reqs, ensures) ->
+     R.seq_ (Lib.list_apply (List.map (check_type_expr ctx)
+                               (List.map (fun (_,_, e) -> e) (reqs @ ensures)))
+               (Bool pos)) 
+  | ContractCall (pos, cname, args, rets) ->
+     R.seq(List.map (infer_type_expr ctx) args)
+     >>= fun ret_tys ->  
+     let ret_ty = if List.length ret_tys = 1
+                  then List.hd ret_tys
+                  else LA.TupleType (pos, ret_tys) in
+     check_type_expr ctx (LA.Call (pos, cname, args))(ret_ty) 
+
+and tc_ctx_of_ty_decl: tc_context -> LA.type_decl -> tc_context tc_result
+  = fun ctx ->
+  function
+  | LA.AliasType (_, i, ty) ->
+     (match ty with
+      | LA.EnumType (pos, n, econsts) ->
+         (match n with
+          | None ->
+             type_error pos "Necessary Enum name not found"
+          | Some ename ->
+             if (List.for_all (fun e -> not (member_ty ctx e)) econsts)
+                && (List.for_all (fun e -> not (member_val ctx e)) econsts)
+             then 
+             R.ok (
+                 List.fold_left union (add_ty_decl ctx ename)
+                   ((List.map ((Lib.flip singleton_ty)
+                                 (LA.UserType (pos, ename))) econsts)
+                    @ Lib.list_apply ((List.map2 (Lib.flip singleton_const)
+                         (List.map (fun i -> LA.Ident (pos, i)) econsts) econsts)) (LA.UserType (pos, ename)))
```

Comment:
This is hard to digest. Let bindings and the `|>` operator are your friends :-)

Reply:
fixed